### PR TITLE
[Operator] slice&select scatter

### PR DESCRIPTION
**Pull request by GwokHiujin**
We have completed the development of the slice_scatter and select_scatter operators. Specifically:

- The corresponding aTen operators are `slice_scatter` and `select_scatter`
- Added accuracy tests and perf tests for slice_scatter and select_scatter

### CODE REVIEW

File: src/flag_gems/utils/shape_utils.py [hunk 1]
```diff
@@ -137,3 +139,30 @@ def dim_compress(inp, dims):
...
     return inp.permute(order).contiguous()
+
+
+def offsetCalculator(inp, idx, strides, dim, isInp):
+    ndim = inp.ndim
+    shape = list(inp.shape)
+    offsets = 0
+    idx_dim = 0
+    for d in range(0, ndim):
+        mod = floor(idx % shape[d])
```
Comment (StrongSpoon) [inline src/flag_gems/utils/shape_utils.py]: I think elements of idx and shape should be integers. Is it necessary to use torch.floor?

File: tests/test_reduction_ops.py [hunk 2]
```diff
@@ -670,3 +670,64 @@ def test_accuracy_vectornorm(shape, ord, dim, keepdim, dtype):
...
+    import random
+
+    index = random.randint(0, shape[dim])
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+
+    src_shape = list(inp.shape)
+    del src_shape[dim]
+    src = torch.randn(src_shape, dtype=dtype, device="cuda")
+
+    ref_inp = to_reference(inp)
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py:460]: src also needs to_reference

File: tests/test_reduction_ops.py [hunk 3]
```diff
@@ -670,3 +670,64 @@ def test_accuracy_vectornorm(shape, ord, dim, keepdim, dtype):
...
+        range = size_dim
+        start = 0
+        end = size_dim
+
+    valid_shape = list(inp.shape)
+    valid_shape[dim] = (range + (step - 1)) // step
+
+    src = torch.randn(valid_shape, dtype=dtype, device="cuda")
+
+    ref_inp = to_reference(inp)
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py:491]: ditto

File: benchmark/test_reduction_perf.py [hunk 4]
```diff
@@ -286,3 +286,69 @@ def test_perf_vector_norm():
...
+def test_perf_slice_scatter():
+    def slice_scatter_args(dtype, batch, size):
+        shape = [batch, size]
+        inp = torch.randn(shape, dtype=dtype, device="cuda")
+        import random
+
+        dim = random.choice([0, 1])
+        start = random.choice([16, 32, 64])
+        end = random.choice([32, 64, 128, 256])
+        step = random.choice([1, 3, 6])
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py]: choosing one of them is enough for benchmark

File: benchmark/test_reduction_perf.py [hunk 5]
```diff
@@ -286,3 +286,69 @@ def test_perf_vector_norm():
...
+def test_perf_select_scatter():
+    def select_scatter_args(dtype, batch, size):
+        shape = [batch, size]
+        inp = torch.randn(shape, dtype=dtype, device="cuda")
+        import random
+
+        dim = random.choice([0, 1])
+        index = random.randint(0, shape[dim])
+        src_shape = shape
+        del src_shape[dim]
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py]: ditto

**Review by StrongSpoon — COMMENTED** [review]
conflicts to resolve

File: tests/test_reduction_ops.py (hunk 2 above)
Comment (GwokHiujin) [reply tests/test_reduction_ops.py:460]: done

File: tests/test_reduction_ops.py (hunk 3 above)
Comment (GwokHiujin) [reply tests/test_reduction_ops.py:491]: done

File: benchmark/test_reduction_perf.py (hunk 4 above)
Comment (GwokHiujin) [reply benchmark/test_reduction_perf.py]: done

File: benchmark/test_reduction_perf.py (hunk 5 above)
Comment (GwokHiujin) [reply benchmark/test_reduction_perf.py]: done

File: src/flag_gems/utils/shape_utils.py (hunk 1 above)
Comment (GwokHiujin) [reply src/flag_gems/utils/shape_utils.py]: I tested it out, and removing torch.floor definitely doesn't affect the correctness. I've gone ahead and removed it.

**Comment by GwokHiujin** [comment]
> conflicts to resolve

solved

File: benchmark/test_reduction_perf.py [hunk 6]
```diff
@@ -286,3 +286,67 @@ def test_perf_vector_norm():
...
+
+        dim = 1
+        start = 32
+        end = 128
+        step = 3
+        size_dim = shape[dim]
+        if start is None:
+            start = 0
+        if end is None:
+            end = size_dim
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py]: remove code from line 301 to line 304

File: benchmark/test_reduction_perf.py [hunk 7]
```diff
@@ -286,3 +286,67 @@ def test_perf_vector_norm():
...
+        range = end - start
+        if end < start:
+            range = 0
+            end = start = 0
+        elif (end - start) > size_dim:
+            range = size_dim
+            start = 0
+            end = size_dim
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py]: remove them too

File: tests/test_reduction_ops.py (hunk 2 above)
Comment (StrongSpoon) [reply tests/test_reduction_ops.py:460]: what I mean is, function to_reference may convert tensor from "cuda" to "cpu" on option --device==cpu, so test as reference should accept tensors returned from to_reference, while test as result only accepts "cuda" tensors.

**Review by StrongSpoon — COMMENTED** [review]
lg. will approve after resolving the conversation in test_reduction_ops.py.

**Comment by GwokHiujin** [comment]
> lg. will approve after resolving the conversation in test_reduction_ops.py.

Already fixed in the latest commit.

**Review by StrongSpoon — DISMISSED** [review]
lgtm

File: benchmark/test_reduction_perf.py [hunk 8]
```diff
@@ -288,6 +288,58 @@ def test_perf_vector_norm():
     bench.run()
 
 
+def test_perf_slice_scatter():
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py]: as we discussed, performance tests are not needed temporarily. we could supply them till the performance is competitive with torch eager.

**Review by StrongSpoon — COMMENTED** [review]
remove perf tests and then we can merge this pr.

File: src/flag_gems/ops/cumsum.py [hunk 9]
```diff
@@ -5,6 +5,8 @@
 import triton
 import triton.language as tl
 
+from ..utils import libentry
```
Comment (StrongSpoon) [inline src/flag_gems/ops/cumsum.py]: need rebasing on master

File: src/flag_gems/utils/shape_utils.py [hunk 10]
```diff
@@ -246,3 +267,24 @@ def offset_calculator(inp, idx, strides, dim, isInp):
             idx_dim = add_on
         idx = idx // shape[d]
     return offsets if not isInp else (offsets - idx_dim)
+
+
+def offsetCalculator(inp, idx, strides, dim, isInp):
+    ndim = inp.ndim
```
Comment (StrongSpoon) [inline src/flag_gems/utils/shape_utils.py:336]: are offset_calculator and offsetCalculator the same? you may update the code and reserve one of them.

File: tests/test_reduction_ops.py [hunk 11]
```diff
@@ -443,9 +443,68 @@ def test_accuracy_gather(inp_shape, dim, dtype):
...
+    ref_src = to_reference(src)
+    ref_out = torch.slice_scatter(
+        ref_inp, dim=dim, src=ref_src, start=start, end=end, step=step
+    )
+    with flag_gems.use_gems():
+        res_out = torch.slice_scatter(
+            ref_inp, dim=dim, src=src, start=start, end=end, step=step
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py]: ref_inp is a cpu tensor and it causes ValueError. use inp instead.

Comment (GwokHiujin) [reply tests/test_reduction_ops.py]: Sorry for the delay in fixing; there was a power outage in the server room. I've made the updates. 

However, I noticed there's still an issue with `test_upsample_bicubic2d_aa` in the unit tests. Should I just ignore it for now?

Comment (StrongSpoon) [reply tests/test_reduction_ops.py]: issue fixed. please rebase on the latest master branch and retry;)

**Review by zhzhcookie — APPROVED** [review]
LGTM

**Pull request merged**